### PR TITLE
add `builder-config` into tarball sources

### PR DESCRIPTION
**Pull request by onur-ozkan**
This will be useful for certain scenarios where developers want to know how the tarball sources were generated. We also want this to check for CI rustc incompatible options on bootstrap.

Blocker for #122709

r? Kobzol

### CODE REVIEW

File: src/bootstrap/src/utils/tarball.rs [hunk 1]
```diff
@@ -317,6 +317,9 @@ impl<'a> Tarball<'a> {
             channel::write_commit_hash_file(&self.overlay_dir, &info.sha);
             channel::write_commit_info_file(&self.overlay_dir, info);
         }
+
+        self.add_file(self.builder.config.config.as_ref().unwrap(), &self.overlay_dir, 0o644);
```
Comment (Kobzol) [inline src/bootstrap/src/utils/tarball.rs:321]: `config.config` can be `None` if you don't specify a path and don't have a created `config.toml`. Probably won't happen on CI, but it shouldn't crash here. I would include the file only if is not `None`.

**Review by Kobzol — COMMENTED** [review]
I think that somehow comparing the config options from CI is a good idea. This is a start, although it won't take into account options configured using the `--set` flags, right?

**Comment by onur-ozkan** [comment]
> This is a start, although it won't take into account options configured using the --set flags, right?

`--set` flag is used with `configure` script which creates the `config.toml`.

**Comment by Kobzol** [comment]
I meant [this](https://github.com/rust-lang/rust/blob/48192701e0385e2f045da79d34bd5e856cf32496/src/bootstrap/src/core/config/flags.rs#L179).

**Comment by onur-ozkan** [comment]
> I meant [this](https://github.com/rust-lang/rust/blob/48192701e0385e2f045da79d34bd5e856cf32496/src/bootstrap/src/core/config/flags.rs#L179).

Oh, yeah that is ignored. I don't think it's important tho.

File: src/bootstrap/src/core/config/config.rs [hunk 2]
```diff
@@ -1325,7 +1325,11 @@ impl Config {
         // Give a hard error if `--config` or `RUST_BOOTSTRAP_CONFIG` are set to a missing path,
         // but not if `config.toml` hasn't been created.
         let mut toml = if !using_default_path || toml_path.exists() {
-            config.config = Some(toml_path.clone());
+            config.config = Some(if cfg!(not(feature = "bootstrap-self-test")) {
```
Comment (Kobzol) [inline src/bootstrap/src/core/config/config.rs:1328]: Maybe we can just use https://doc.rust-lang.org/stable/std/path/fn.absolute.html instead? Unless you also wanted to check here that the path exists.

**Review by Kobzol — COMMENTED** [review]
I don't think that we should add the config file to *all* the tarballs, even those that we publicly distribute to users. It is an implementation detail, and more importantly it makes it harder to reproduce the archives, especially the source code archive - that should only contain things that are in git, so that it can be perfectly reproduced to have the same SHA.

Could we only add it to the internal component(s) that need it, so for now `RustcDev`?

File: src/bootstrap/src/core/config/config.rs (hunk 2 above)
Comment (onur-ozkan) [reply src/bootstrap/src/core/config/config.rs:1328]: `canonicalize` also helps to resolve symlinks.

Comment (Kobzol) [reply src/bootstrap/src/core/config/config.rs:1328]: Good point, let's keep it then.

**Comment by Kobzol** [comment]
Clarification: the config file isn't added to bare tarballs (such as the source code archive), so it should be fine to add it unconditionally.

@bors r+

**Pull request merged**